Propose a blog post on some performance fixes contributions (#1520)
opensearch-project/project-website: _posts/2023-04-10-opensource-perf.markdown

---
layout: post
title:  "Improving JSON parsing performance in opensearch-java"
authors:
  - chibenwa
date:   2023-04-10
categories:
  - technical-post
meta_keywords: OpenSearch, Apache James, Flame graph, Performance, Driver

> meta_description: Learn how contributors of Apache James found out and fixed major performance problems in OpenSearch Java driver JSON parsing.

Reviewer: Please use the following meta for the description:

**meta_description:** Learn how Apache James project contributors identified and assessed issues with the OpenSearch Java driver to improve JSON parsing performance.

---

> As an open source enthusiast, I believe in the power of collaboration to make open source projects faster and more efficient. In this blog post, I will share how my team at [LINAGORA](https://linagora.com), contributing to the Apache James project, identified and addressed a performance issue in the OpenSearch java client using benchmark tools and flamegraphs, and collaborated with the OpenSearch community.

Author: ```suggestion
As an open source enthusiast, I believe in the power of collaboration to make open source projects faster and more efficient. In this blog post, I will share how my team at [Linagora](https://linagora.com), contributing to the Apache James project, identified and addressed a performance issue in the OpenSearch java client using benchmark tools and flamegraphs, and collaborated with the OpenSearch community.
```

> As an open source enthusiast, I believe in the power of collaboration to make open source projects faster and more efficient. In this blog post, I will share how my team at [LINAGORA](https://linagora.com), contributing to the Apache James project, identified and addressed a performance issue in OpenSearch using benchmark tools and flamegraphs, and collaborated with the OpenSearch community to make the project better.

Reviewer 2: performance issue in **the OpenSearch Java client**

> As an open source enthusiast, I believe in the power of collaboration to make open source projects faster and more efficient. In this blog post, I will share how my team at [LINAGORA](https://linagora.com), contributing to the Apache James project, identified and addressed a performance issue in OpenSearch using benchmark tools and flamegraphs, and collaborated with the OpenSearch community to make the project better.

Reviewer 2: I would remove "to make the project better" because it dilutes the "identified and addressed a performance issue"

> As an open source enthusiast, I believe in the power of collaboration to make open source projects faster and more efficient. In this blog post, I will share how my team at [LINAGORA](https://linagora.com), contributing to the Apache James project, identified and addressed a performance issue in OpenSearch using benchmark tools and flamegraphs, and collaborated with the OpenSearch community to make the project better.

Author: ```suggestion
As an open source enthusiast, I believe in the power of collaboration to make open source projects faster and more efficient. In this blog post, I will share how my team at [LINAGORA](https://linagora.com), contributing to the Apache James project, identified and addressed a performance issue in the OpenSearch java client using benchmark tools and flamegraphs, and collaborated with the OpenSearch community.
```

When operating an email server, performance is a concern, as email is a massively used application, generally considered critical. Often with even a medium-sized deployment, handling over 1000 requests per seconds is not uncommon. We use a range of tools for benchmarking the performance of Apache James. Our docs [explain](https://github.com/apache/james-project/blob/master/server/apps/distributed-app/docs/modules/ROOT/pages/benchmark/index.adoc) some of the benchmarks we suggest our administrators run to identify performance bottlenecks and malfunctions in their clusters.

> Apache James relies on OpenSearch as its search engine for its email database. It takes as a dependency the [opensearch-java client](https://github.com/opensearch-project/opensearch-java). We frequently run performance tests with custom [Gatling](https://gatling.io/) benchmarks. We specifically wanted to regression test performance when we migrated from ElasticSearch 7.10 to OpenSearch for licenses reasons (as an Apache project, the SSPL license of driver code is not acceptable). We realized the OpenSearch requests were slower and started investigating why.

Reviewer 2: Maybe soften "SSPL license of driver code is not acceptable" to "we are not permitted to use non-open-source code" or "not allowed to depend on non-APLv2 code"?

> Apache James relies on OpenSearch as its search engine for its email database. It takes as a dependency the [opensearch-java client](https://github.com/opensearch-project/opensearch-java). We frequently run performance tests with custom [Gatling](https://gatling.io/) benchmarks. We specifically wanted to regression test performance when we migrated from ElasticSearch 7.10 to OpenSearch for licenses reasons (as an Apache project, the SSPL license of driver code is not acceptable). We realized the OpenSearch requests were slower and started investigating why.

Author: ```suggestion
Apache James relies on OpenSearch as its search engine for its email database. It takes as a dependency the [opensearch-java client](https://github.com/opensearch-project/opensearch-java). We frequently run performance tests with custom [Gatling](https://gatling.io/) benchmarks. We specifically wanted to regression test performance when we migrated from ElasticSearch 7.10 to OpenSearch for licenses reasons (as an Apache project, we must use licenses compatible with the Apache License 2.0). We realized the OpenSearch requests were slower and started investigating why.
```

### Identifying the performance issue

Flamegraphs are powerful visualization tools that can help developers analyze the execution flow of their code and identify bottlenecks. By analyzing the flamegraphs, I was able to identify specific areas of code that were causing the performance issue. We used async-profiler to generate a flamegraph and found that frequent SPI lookups were causing the performance issue.

> [opensearch_2.4.0.zip](https://github.com/opensearch-project/opensearch-java/files/10334079/opensearch_2.4.0.zip) is the flame graph we diagnosed this issue from. Taken with <https://github.com/jvm-profiling-tools/async-profiler> (From within my applicative container):

Reviewer 2: Make it a full sentence? You can download the flame graph taken with ... in my container ...

> [opensearch_2.4.0.zip](https://github.com/opensearch-project/opensearch-java/files/10334079/opensearch_2.4.0.zip) is the flame graph we diagnosed this issue from. Taken with <https://github.com/jvm-profiling-tools/async-profiler> (From within my applicative container):

Author: ```suggestion
[opensearch_2.4.0.zip](https://github.com/opensearch-project/opensearch-java/files/10334079/opensearch_2.4.0.zip) is the flame graph we diagnosed this issue from. It was taken with <https://github.com/jvm-profiling-tools/async-profiler>  from within Apache James container):
```

> [opensearch_2.4.0.zip](https://github.com/opensearch-project/opensearch-java/files/10334079/opensearch_2.4.0.zip) is the flame graph we diagnosed this issue from. It was taken with <https://github.com/jvm-profiling-tools/async-profiler>  from within Apache James container):

Reviewer 2: I would move the link to the graph below the graph, and maybe make it a full sentence, e.g. `You can download the interactive flame graph that was used to diagnose this issue [here](...).`

> [opensearch_2.4.0.zip](https://github.com/opensearch-project/opensearch-java/files/10334079/opensearch_2.4.0.zip) is the flame graph we diagnosed this issue from. It was taken with <https://github.com/jvm-profiling-tools/async-profiler>  from within Apache James container):

Reviewer 2: Does the link https://github.com/jvm-profiling-tools/async-profiler work? Maybe just make it `[async-profiler](https://github.com/jvm-profiling-tools/async-profiler)`

> [opensearch_2.4.0.zip](https://github.com/opensearch-project/opensearch-java/files/10334079/opensearch_2.4.0.zip) is the flame graph we diagnosed this issue from. It was taken with <https://github.com/jvm-profiling-tools/async-profiler>  from within Apache James container):

Author: ```suggestion
Here is the flame graph we diagnosed this issue from. It was taken with [async-profiler](https://github.com/jvm-profiling-tools/async-profiler)  from within Apache James container:
```

`./profiler -d 120 -e itimer -f opensearch_2.4.0.html 1`

<img src="/assets/media/blog-images/2023-04-10-opensource-perf/flame1.png" alt="Flame graph general overview"/>{: .img-fluid }

> Then searching for opensearch using the search widget in the top left corner of the flame graph:

Reviewer 2: opensearch -> quote it since we're talking a string and not the product, "opensearch"

> Then searching for opensearch using the search widget in the top left corner of the flame graph:

Reviewer 2: Is it "flamegraph" or "flame graph"? Use the same in the entire post.

> Then searching for opensearch using the search widget in the top left corner of the flame graph:

Author: ```suggestion
Then searching for `"opensearch"` using the search widget in the top left corner of the flamegraph:
```

<img src="/assets/media/blog-images/2023-04-10-opensource-perf/flame2.png" alt="Flame graph: OpenSearch driver threads"/>{: .img-fluid }

> -   The driver thread (on the right) with the httpclient event loop

Reviewer 2: If it's a full sentence finish with a period, or lowercase the first word

> -   The driver thread (on the right) with the httpclient event loop

Author: ```suggestion
-   The driver thread (on the right) with the httpclient event loop.
```

-   Apache James calls to the OpenSearch driver (small stacktraces in the center)

Observations on the driver event loop:

> -   We spend a huge amount of resources doing SPI calls in order to locate the JSON parser implementation. It not only eats CPU/heap allocations but it also blocks the HTTP event loop, which can be catastrophic (the James app is not OpenSearch heavy so that is not a big issue to me).

Reviewer 2: huge amount -> something more specific like a %?

> -   We spend a huge amount of resources doing SPI calls in order to locate the JSON parser implementation. It not only eats CPU/heap allocations but it also blocks the HTTP event loop, which can be catastrophic (the James app is not OpenSearch heavy so that is not a big issue to me).

Author: ```suggestion
-   We spend a huge amount of resources doing SPI calls in order to locate the JSON parser implementation. It not only eats CPU (roughly 26% of driver CPU consumption) /heap allocations.
-   But it also blocks the HTTP event loop, which can be catastrophic: the driver relies on a few thread submitting requests in parallel thus blocking the thread impacts seriously overall latencies and throughput.
```

-   Unsurprisingly JSON parsing takes 11% of the driver's CPU and 24% of heap allocation.
-   Event loop busyness (pushing stuff to the kernel network stack, SSL) takes around 60% of the event loop CPU, 25% of heap allocations, which again feels normal to me.
-   2.3% of the event loop CPU is our binding to Reactor reactive library, which, again is normal: converting futures and enqueuing tasks takes time.

> To mitigate calls to the SPI, we submitted a [small change](https://github.com/opensearch-project/opensearch-java/pull/293/files) to JsonValueParser.java that addressed the issue. This led to a 50x speedup of the JSON parsing by getting rid of the SPI lookups, not even taking into account the blocking operations... Huge win!

Reviewer 2: small change -> a change

remove /files from the PR URL

link JsonValueParser.java

> To mitigate calls to the SPI, we submitted a [small change](https://github.com/opensearch-project/opensearch-java/pull/293/files) to JsonValueParser.java that addressed the issue. This led to a 50x speedup of the JSON parsing by getting rid of the SPI lookups, not even taking into account the blocking operations... Huge win!

Author: ```suggestion
To mitigate calls to the SPI, we submitted a [change](https://github.com/opensearch-project/opensearch-java/pull/293) to [JsonValueParser.java](https://github.com/opensearch-project/opensearch-java/blob/a8df7e7c26ccc644811539c4fea57d97f1031aaa/java-client/src/main/java/org/opensearch/client/json/jackson/JsonValueParser.java#L52) that addressed the issue. This led to a 50x speedup of the JSON parsing by getting rid of the SPI lookups, not even taking into account the blocking operations... Huge win!
```

> 

Author: ```suggestion

You can download the interactive flame graph that was used to diagnose this issue [here](https://github.com/opensearch-project/opensearch-java/files/10334079/opensearch_2.4.0.zip).

```

As a common practice we then run micro-benchmarks to validate changes, and [JMH](https://github.com/openjdk/jmh) comes in handy for this! It summarizes key metrics, performs warmup, repeats measurements and comes with nano-second resolution!

> Here is the JMH output backing this change:

Reviewer 2: Include how to run the benchmark?

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.ObjectMapper;

public class JMHFieldBench {
    public static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    @Test
    public void launchBenchmark() throws Exception {
        Options opt = new OptionsBuilder()
            .include(this.getClass().getName() + ".*")
            .mode (Mode.AverageTime)
            .timeUnit(TimeUnit.MICROSECONDS)
            .warmupTime(TimeValue.seconds(5))
            .warmupIterations(3)
            .measurementTime(TimeValue.seconds(2))
            .measurementIterations(5)
            .threads(1)
            .forks(1)
            .shouldFailOnError(true)
            .shouldDoGC(true)
            .build();
        new Runner(opt).run();
    }

    @Benchmark
    public void jsonBench(Blackhole bh) throws IOException {
        final JsonParser parser = OBJECT_MAPPER.createParser("[\"a\",\"b\"]");
        try {
            bh.consume(new JacksonJsonpParser(parser).getArrayStream());
        } catch (Exception e) {
            // do nothing
        } finally {
            parser.close();
        }
    }
}
```

> The bench was applied before/after the backport of this PR on top of Apache James CF [apache/james-project@d5af3a5](https://github.com/apache/james-project/commit/d5af3a52cd30eebf7a8fb4d8f2402920c42d5f7c)

Reviewer 2: Add a perioid.

---
layout: post

> title:  "Collaborating to Improve Open Source Performance: A Case Study"

Reviewer 2: I appreciate your calling out our collaboration, but you did all the work! I would just call it "Improving JSON parsing performance in opensearch-java" and removing this "dblock" person from the authors!

> title:  "Collaborating to Improve Open Source Performance: A Case Study"

Author: ```suggestion
title:  "Improving JSON parsing performance in opensearch-java"
```

authors:

>   - wbeckler

Author: As you wish

>   - dblock
>   - chibenwa

Author: ```suggestion
  - wbeckler
  - chibenwa
```

date:   2023-04-10
categories:
  - technical-post
meta_keywords: OpenSearch, Apache James, Flame graph, Performance, Driver

> meta_description: Learn how key contributors of APache James found out and fixed major performance problems in OpenSearch Java driver.

Reviewer 2: Apache lowercase P.
Drop "key".
Get specific on what the performance problem was (JSON parsing?)

> meta_description: Learn how key contributors of APache James found out and fixed major performance problems in OpenSearch Java driver.

Author: ```suggestion
meta_description: Learn how contributors of Apache James found out and fixed major performance problems in OpenSearch Java driver JSON parsing.
```

Operating an email server, performance is a concern, as emails are massively used applications, generaly considered critical. Often with even medium sized deployment, handling over 1000 requests per seconds is not uncommon. We use a range of tools for benchmarking the performance of Apache James. Our docs [explains](https://github.com/apache/james-project/blob/master/server/apps/distributed-app/docs/modules/ROOT/pages/benchmark/index.adoc) some of the benchmarks we suggest our administrators run to identify performance bottlenecks and malfunctions in their clusters.

> Apache James relies on OpenSearch as its search engine for its email database. It takes as a dependency the [opensearch-java client](https://github.com/opensearch-project/opensearch-java). Performance testing revealed that the application was spending a noticeable chunk of time running the opensearch-java code, so we set out to examine the performance characteristics of this component.

Reviewer 3: Could you maybe clarify how this was detected? As in, which particular benchmark or analysis revealed this disparity?

> Apache James relies on OpenSearch as its search engine for its email database. It takes as a dependency the [opensearch-java client](https://github.com/opensearch-project/opensearch-java). Performance testing revealed that the application was spending a noticeable chunk of time running the opensearch-java code, so we set out to examine the performance characteristics of this component.

Author: ```suggestion
Apache James relies on OpenSearch as its search engine for its email database. It takes as a dependency the [opensearch-java client](https://github.com/opensearch-project/opensearch-java). We frequently run performance tests with custom [Gatling](https://gatling.io/) made benchmarks. Especially we conducting regression tests on performance when migrating from ElasticSearch 7.10 to OpenSearch for licenses reasons (as an Apache project the SSPL license of driver code is not acceptable). We realized the OpenSearch requests were slower and started investigating why.
```

### Identifying the performance issue

> Flamegraphs are powerful visualization tools that can help developers analyze the execution flow of their code and identify bottlenecks. By analyzing the flamegraphs, I was able to identify specific areas of code that were causing the performance issue. We used async-profiler to generate a flamegraph and found that the frequent SPI lookups were causing the performance issue.

Reviewer 3: ```suggestion
Flamegraphs are powerful visualization tools that can help developers analyze the execution flow of their code and identify bottlenecks. By analyzing the flamegraphs, I was able to identify specific areas of code that were causing the performance issue. We used async-profiler to generate a flamegraph and found that frequent SPI lookups were causing the performance issue.
```

[opensearch_2.4.0.zip](https://github.com/opensearch-project/opensearch-java/files/10334079/opensearch_2.4.0.zip) is the flame graph we diagnosed this issue from. Taken with <https://github.com/jvm-profiling-tools/async-profiler> (From within my applicative container):

`./profiler -d 120 -e itimer -f opensearch_2.4.0.html 1`

<img src="/assets/media/blog-images/2023-04-10-opensource-perf/flame1.png" alt="Flame graph general overview"/>{: .img-fluid }

Then searching for opensearch using the search widget in the top left corner of the flame graph:

<img src="/assets/media/blog-images/2023-04-10-opensource-perf/flame2.png" alt="Flame graph: OpenSearch driver threads"/>{: .img-fluid }

-   The driver thread (on the right) with the httpclient event loop

> -   Applicatives calls to the driver from applicative threads (small stacktraces in the center)

Author: ```suggestion
-   Apache James calls to the OpenSearch driver (small stacktraces in the center)
```

> -   Applicatives calls to the driver from applicative threads (small stacktraces in the center)

Reviewer 3: I don't think "applicative" is the right word in English, but I'm not sure what is the right word to use.

Observations on the driver event loop:

-   Calls to the SPI make it easy to notice a CF issue. It not only eats CPU/heap allocations but it also blocks the HTTP event loop, which can be catastrophic (the James app is not OpenSearch heavy so that is not a big issue to me).

> -   Unsurprisingly JSON parsing takes 11% of the drivers CPU and 24% of heap allocation.

Reviewer 3: ```suggestion
-   Unsurprisingly JSON parsing takes 11% of the driver's CPU and 24% of heap allocation.
```

-   Event loop busyness (pushing stuff to the kernel network stack, SSL) takes around 60% of the event loop CPU, 25% of heap allocations, which again feels normal to me.
-   2.3% of the event loop CPU is our binding to Reactor reactive library, which, again is normal: converting futures and enqueuing tasks takes time.

To mitigate calls to the SPI, we submitted a [small change](https://github.com/opensearch-project/opensearch-java/pull/293/files) to JsonValueParser.java that addressed the issue. This led to a 50x speedup of the JSON parsing by getting rid of the SPI lookups, not even taking into account the blocking operations... Huge win!

> As a common practise we then run micro-benchmarks to validate changes, and [JMH](https://github.com/openjdk/jmh) comes in handy for this! It summarizes key metrics, performs warmup, repeats measurement and comes with nano-second resulution!

Reviewer 3: ```suggestion
As a common practice we then run micro-benchmarks to validate changes, and [JMH](https://github.com/openjdk/jmh) comes in handy for this! It summarizes key metrics, performs warmup, repeats measurements and comes with nano-second resolution!
```

---
layout: post
title:  "Improving JSON parsing performance in opensearch-java"
authors:
  - chibenwa
date:   2023-04-10
categories:
  - technical-post

> meta_keywords: OpenSearch, Apache James, Flame graph, Performance, Driver

Reviewer: Keywords are too broad, please update the meta with the following: 

**meta_keywords:** Apache James, flame graph, flamegraph, JSON parsing performance< OpenSearch Java driver

When operating an email server, performance is a concern, as email is a massively used application, generally considered critical. Often with even a medium-sized deployment, handling over 1000 requests per seconds is not uncommon. We use a range of tools for benchmarking the performance of Apache James. Our docs [explain](https://github.com/apache/james-project/blob/master/server/apps/distributed-app/docs/modules/ROOT/pages/benchmark/index.adoc) some of the benchmarks we suggest our administrators run to identify performance bottlenecks and malfunctions in their clusters.

Apache James relies on OpenSearch as its search engine for its email database. It takes as a dependency the [opensearch-java client](https://github.com/opensearch-project/opensearch-java). We frequently run performance tests with custom [Gatling](https://gatling.io/) benchmarks. We specifically wanted to regression test performance when we migrated from ElasticSearch 7.10 to OpenSearch for licenses reasons (as an Apache project, we must use licenses compatible with the Apache License 2.0). We realized the OpenSearch requests were slower and started investigating why.

> ### Identifying the performance issue

Reviewer 4: ```suggestion
## Identifying the performance issue
```

> Flamegraphs are powerful visualization tools that can help developers analyze the execution flow of their code and identify bottlenecks. By analyzing the flamegraphs, I was able to identify specific areas of code that were causing the performance issue. We used async-profiler to generate a flamegraph and found that frequent SPI lookups were causing the performance issue.

Reviewer 4: Should we define SPI for readers who may not be familiar?

> Flamegraphs are powerful visualization tools that can help developers analyze the execution flow of their code and identify bottlenecks. By analyzing the flamegraphs, I was able to identify specific areas of code that were causing the performance issue. We used async-profiler to generate a flamegraph and found that frequent SPI lookups were causing the performance issue.

Reviewer 4: ```suggestion
Flame graphs are powerful visualization tools that can help developers analyze the execution flow of their code and identify bottlenecks. By analyzing the flame graphs, we were able to identify specific areas of code that were causing the performance issue. We used async-profiler to generate a flame graph and found that frequent SPI lookups were causing the performance issue.
```

`./profiler -d 120 -e itimer -f opensearch_2.4.0.html 1`

<img src="/assets/media/blog-images/2023-04-10-opensource-perf/flame1.png" alt="Flame graph general overview"/>{: .img-fluid }

Then searching for `"opensearch"` using the search widget in the top left corner of the flamegraph:

<img src="/assets/media/blog-images/2023-04-10-opensource-perf/flame2.png" alt="Flame graph: OpenSearch driver threads"/>{: .img-fluid }

-   the driver thread (on the right) with the httpclient event loop
-   Apache James calls to the OpenSearch driver (small stacktraces in the center)

Observations on the driver event loop:

-   We spend a huge amount of resources doing SPI calls in order to locate the JSON parser implementation. It not only eats CPU (roughly 26% of driver CPU consumption) /heap allocations.
-   But it also blocks the HTTP event loop, which can be catastrophic: the driver relies on a few thread submitting requests in parallel thus blocking the thread impacts seriously overall latencies and throughput.

> -   Unsurprisingly JSON parsing takes 11% of the driver's CPU and 24% of heap allocation.

Reviewer 4: ```suggestion
-   Unsurprisingly, JSON parsing takes 11% of the driver's CPU and 24% of its heap allocation.
```

> -   Event loop busyness (pushing stuff to the kernel network stack, SSL) takes around 60% of the event loop CPU, 25% of heap allocations, which again feels normal to me.

Reviewer 4: ```suggestion
-   Event loop busyness (pushing stuff to the kernel network stack, SSL) takes around 60% of the event loop CPU and 25% of heap allocations, which appears normal.
```

> -   Event loop busyness (pushing stuff to the kernel network stack, SSL) takes around 60% of the event loop CPU, 25% of heap allocations, which again feels normal to me.

Reviewer 4: I would use a word other than "stuff." Can we be more precise?

> -   2.3% of the event loop CPU is our binding to Reactor reactive library, which, again is normal: converting futures and enqueuing tasks takes time.

Reviewer 4: ```suggestion
-   2.3% of the event loop CPU is our binding to a Reactor reactive library, which, again, is normal: converting futures and enqueuing tasks takes time.
```

> To mitigate calls to the SPI, we submitted a [change](https://github.com/opensearch-project/opensearch-java/pull/293) to [JsonValueParser.java](https://github.com/opensearch-project/opensearch-java/blob/a8df7e7c26ccc644811539c4fea57d97f1031aaa/java-client/src/main/java/org/opensearch/client/json/jackson/JsonValueParser.java#L52) that addressed the issue. This led to a 50x speedup of the JSON parsing by getting rid of the SPI lookups, not even taking into account the blocking operations... Huge win!

Reviewer 4: ```suggestion
To mitigate calls to the SPI, we submitted a [change](https://github.com/opensearch-project/opensearch-java/pull/293) to [JsonValueParser.java](https://github.com/opensearch-project/opensearch-java/blob/a8df7e7c26ccc644811539c4fea57d97f1031aaa/java-client/src/main/java/org/opensearch/client/json/jackson/JsonValueParser.java#L52) that addressed the issue. This led to a 50x speedup of the JSON parsing by getting rid of the SPI lookups, not even taking into account the blocking operations...huge win!
```

You can download the interactive flame graph that was used to diagnose this issue [here](https://github.com/opensearch-project/opensearch-java/files/10334079/opensearch_2.4.0.zip).

> As a common practice we then run micro-benchmarks to validate changes, and [JMH](https://github.com/openjdk/jmh) comes in handy for this! It summarizes key metrics, performs warmup, repeats measurements and comes with nano-second resolution!

Reviewer 4: ```suggestion
As a common practice, we then run micro-benchmarks to validate changes, and [JMH](https://github.com/openjdk/jmh) comes in handy for this. It summarizes key metrics, performs warmup, repeats measurements, and comes with nanosecond resolution!
```

> Here is the JMH output backing this change:

Reviewer 4: ```suggestion
Here is the JMH output backing this change.
```

#### Before

```
Benchmark                Mode  Cnt   Score   Error  Units
JMHFieldBench.jsonBench  avgt    5  71.790 ± 2.125  us/op
```

#### After

```
Benchmark                Mode  Cnt  Score   Error  Units
JMHFieldBench.jsonBench  avgt    5  1.418 ± 0.024  us/op
```

> #### Code of the benchmark

Reviewer 4: "Benchmark code"?

    @Test
    public void launchBenchmark() throws Exception {
        Options opt = new OptionsBuilder()
            .include(this.getClass().getName() + ".*")
            .mode (Mode.AverageTime)
            .timeUnit(TimeUnit.MICROSECONDS)
            .warmupTime(TimeValue.seconds(5))
            .warmupIterations(3)
            .measurementTime(TimeValue.seconds(2))
            .measurementIterations(5)
            .threads(1)
            .forks(1)
            .shouldFailOnError(true)
            .shouldDoGC(true)
            .build();
        new Runner(opt).run();
    }

    @Benchmark
    public void jsonBench(Blackhole bh) throws IOException {
        final JsonParser parser = OBJECT_MAPPER.createParser("[\"a\",\"b\"]");
        try {
            bh.consume(new JacksonJsonpParser(parser).getArrayStream());
        } catch (Exception e) {
            // do nothing
        } finally {
            parser.close();
        }
    }
}
```

The bench was applied before/after the backport of this PR on top of Apache James CF [apache/james-project@d5af3a5](https://github.com/apache/james-project/commit/d5af3a52cd30eebf7a8fb4d8f2402920c42d5f7c).

> ### Collaborating with the OpenSearch community

Reviewer 4: ```suggestion
## Collaborating with the OpenSearch community
```

> I believe open source projects are more than just code--they're communities of people who share a common goal of making technology more accessible and efficient. By collaborating with the OpenSearch community, I contributed to this goal by getting rid of SPI lookups and improving JSON parsing performance in an application that is probably running on countless servers. Working together, we made OpenSearch faster and more efficient. It's rewarding to make a positive impact on both technology and the community of people who use it. This case study shows how collaboration and community involvement can make open source technology better for everyone.

Reviewer 4: ```suggestion
I believe open-source projects are more than just code---they're communities of people who share a common goal of making technology more accessible and efficient. By collaborating with the OpenSearch community, we contributed to this goal by getting rid of SPI lookups and improving JSON parsing performance in an application that is probably running on countless servers. Working together, we made OpenSearch faster and more efficient. It's rewarding to make a positive impact on both technology and the community of people who use it. This case study shows how collaboration and community involvement can make open-source technology better for everyone.
```
